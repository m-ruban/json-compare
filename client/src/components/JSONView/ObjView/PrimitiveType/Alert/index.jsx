import React from 'react';
import PropTypes from 'prop-types';
import { COMPARE_REQ, COMPARE_DIFF } from 'constants/compare';
import Mark from 'components/Mark';

const Alert = ({ alert, showAlerts, showRequired }) => {
    let node = null;
    if ((showAlerts || showRequired) && alert) {
        let show = false;
        if (showAlerts) show = alert.res === COMPARE_DIFF;
        if (showRequired && !show) show = alert.res === COMPARE_REQ;

        if (show) {
            node = <Mark type={alert.res} text="!" pill />;
        }
    }

    return node;
};

Alert.propTypes = {
    alert: PropTypes.object,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
};

export default Alert;
