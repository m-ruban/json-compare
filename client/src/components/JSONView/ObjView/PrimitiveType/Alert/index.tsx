import React from 'react';

import Mark from 'components/Mark';
import { COMPARE_DIFF, COMPARE_REQ } from 'constants/compare';

interface IAlert {
    res: string;
    scalar: boolean;
}

interface IAlertProps {
    alert: IAlert;
    showAlerts: boolean;
    showRequired: boolean;
}

const Alert = ({ alert, showAlerts, showRequired }: IAlertProps) => {
    let node = null;
    if ((showAlerts || showRequired) && alert) {
        let show = false;
        if (showAlerts) { show = alert.res === COMPARE_DIFF; }
        if (showRequired && !show) { show = alert.res === COMPARE_REQ; }

        if (show) {
            node = <Mark type={alert.res} text="!" pill />;
        }
    }

    return node;
};

export default Alert;
