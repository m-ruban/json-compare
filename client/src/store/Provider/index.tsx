import * as React from 'react';

import { COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT } from 'constants/equality';
import { AUTHOR_LINK, FEEDBACK_LINK } from 'constants/links';
import { LOG_INFO } from 'constants/log';
import { Provider } from 'store/Context';

import { equalAlerts } from 'modules/compare';
import { parseAndLog } from 'modules/parse';

interface IProviderProps {
    children: any;
}

interface IProviderState {
    alerts: Map<string, object>;
    author: string;
    changeEqualityTypes: (event: React.SyntheticEvent) => void;
    changeObj: (event: React.SyntheticEvent) => void;
    changeTool: (value: string) => void;
    equalityTypes: string;
    feedback: string;
    isCompare: boolean;
    log: object;
    objObject: object;
    objStr: string;
    toggleCompare: (event: React.SyntheticEvent) => void;
    toolObject: object;
    toolStr: string;
}

class AppProvider extends React.Component<IProviderProps, IProviderState> {
    constructor(props) {
        super(props);
        this.state = {
            alerts: new Map(),
            author: AUTHOR_LINK,
            changeEqualityTypes: this.changeEqualityTypes.bind(this),
            changeObj: this.changeObj.bind(this),
            changeTool: this.changeTool.bind(this),
            equalityTypes: EQUALITY_TYPE_STRICT,
            feedback: FEEDBACK_LINK,
            isCompare: false,
            log: { text: '', type: null },
            objObject: {},
            objStr: '',
            toggleCompare: this.toggleCompare.bind(this),
            toolObject: {},
            toolStr: '',
        };
    }

    public toggleCompare() {
        const log = {
            text: 'toggle compare',
            type: LOG_INFO,
        };

        this.setState({ isCompare: !this.state.isCompare, log }, this.changeAlerts);
    }

    public changeEqualityTypes(event) {
        const { value } = event.target;
        const log = {
            text: 'change equality types',
            type: LOG_INFO,
        };
        this.setState({ equalityTypes: value, log }, this.changeAlerts);
    }

    public changeTool(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ toolObject: obj, toolStr: value, log }, this.changeAlerts);
    }

    public changeObj(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ objObject: obj, objStr: value, log }, this.changeAlerts);
    }

    public changeAlerts() {
        const { toolObject, objObject, equalityTypes, isCompare } = this.state;
        const isEmpty = (obj) => Object.keys(obj).length === 0;

        if (!isEmpty(toolObject) && !isEmpty(objObject) && isCompare) {
            const alerts = equalAlerts(toolObject, objObject, equalityTypes);
            let req = 0;
            let diff = 0;
            let text = '';

            alerts.forEach((alert) => {
                if (!alert.scalar) { return; }
                if (alert.res === COMPARE_REQ) {
                    req++;
                } else {
                    diff++;
                }
            });

            if (diff > 0 || req > 0) {
                text = `${diff} alert${diff > 1 ? 's' : ''}, ${req} required`;
            } else {
                text = 'objects are equal';
            }

            this.setState({
                alerts,
                log: { text, type: LOG_INFO },
            });
        }
    }

    public render() {
        return <Provider value={this.state}>{this.props.children}</Provider>;
    }
}

export default AppProvider;
