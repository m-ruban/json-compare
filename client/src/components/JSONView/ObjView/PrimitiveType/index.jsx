import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import withConsumer from 'hoc/withConsumer';
import { STRING_TYPE } from 'constants/types';
import Alert from 'components/JSONView/ObjView/PrimitiveType/Alert';

const PrimitiveType = ({ value, showAlerts, showRequired, path, alerts }) => {
    const mark = typeof value === STRING_TYPE ? '"' : null;
    const key = path.join('-');
    const rValue = value !== null ? value.toString() : 'null';
    return (
        <Fragment>
            {mark}
            {rValue}
            {mark}
            <Alert id={key} alert={alerts.get(key)} showAlerts={showAlerts} showRequired={showRequired} />
        </Fragment>
    );
};

PrimitiveType.propTypes = {
    value: PropTypes.oneOfType([PropTypes.bool, PropTypes.string, PropTypes.number]),
    path: PropTypes.array,
    showAlerts: PropTypes.bool,
    showRequired: PropTypes.bool,
    alerts: PropTypes.instanceOf(Map),
};

export default withConsumer(PrimitiveType);
