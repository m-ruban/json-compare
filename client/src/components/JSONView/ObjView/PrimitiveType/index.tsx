import React, { Fragment } from 'react';

import Alert from 'components/JSONView/ObjView/PrimitiveType/Alert';
import { IAlertMap } from 'constants/interfaces';
import { STRING_TYPE } from 'constants/types';
import withConsumer from 'hoc/withConsumer';

interface IPrimitiveTypeProps extends IAlertMap {
    value: boolean | string | number;
    showAlerts: boolean;
    showRequired: boolean;
    path: string[];
}

const PrimitiveType = ({ value, showAlerts, showRequired, path, alerts }: IPrimitiveTypeProps) => {
    const mark = typeof value === STRING_TYPE ? '"' : null;
    const key = path.join('-');
    const rValue = value !== null ? value.toString() : 'null';
    return (
        <span>
            {mark}
            {rValue}
            {mark}
            <Alert alert={alerts.get(key)} showAlerts={showAlerts} showRequired={showRequired} />
        </span>
    );
};

export default withConsumer(PrimitiveType);
