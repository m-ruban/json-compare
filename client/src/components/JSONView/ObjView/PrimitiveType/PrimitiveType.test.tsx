import React from 'react';

import PrimitiveType from 'components/JSONView/ObjView/PrimitiveType';
import { render } from 'enzyme';
import { Provider } from 'store/Context';
import alerts from 'tests/mocks/alerts';

describe('<PrimitiveType />', () => {
    it('Renders correctly', () => {
        const props = { value: 'test', showAlerts: false, showRequired: false, path: [] };
        const output = render(
            <Provider value={alerts}>
                <PrimitiveType {...props} />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
