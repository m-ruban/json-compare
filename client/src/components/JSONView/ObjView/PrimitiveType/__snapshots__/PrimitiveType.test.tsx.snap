// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrimitiveType /> Renders correctly 1`] = `
<span>
  "test"
</span>
`;
